particles/Tutorial/Kotlin/Demo/TTTGame.kt

```diff
@@ -0,0 +1,173 @@
+    private val playerTwo = Singleton { TTTGame_PlayerTwo() }
+    private val playerTwoMove = Singleton { TTTGame_PlayerTwoMove() }
+    private val events = Collection { TTTGame_Events() }
+
+    private val winningSequences = arrayOf(
+        arrayOf(0, 1, 2),
+        arrayOf(3, 4, 5),
+        arrayOf(6, 7, 8),
+        arrayOf(0, 3, 6),
+        arrayOf(1, 4, 7),
+        arrayOf(2, 5, 8),
+        arrayOf(0, 4, 8),
+        arrayOf(2, 4, 6)
+    )
+
+    private val defaultGame = TTTGame_GameState(
+        board = ",,,,,,,,",
+        currentPlayer = (0..1).random().toDouble(),
+        gameOver = false
+    )
+
+    init {
+        registerHandle("gameState", gameState)
+        registerHandle("playerOne", playerOne)
+        registerHandle("playerOneMove", playerOneMove)
+        registerHandle("playerTwo", playerTwo)
+        registerHandle("playerTwoMove", playerTwoMove)
+        registerHandle("events", events)
+    }
+
+    override fun onHandleSync(handle: Handle, allSynced: Boolean) {
+        if (gameState.get()?.board == null) {
+            gameState.set(defaultGame)
+        }
+        if (handle.name.equals("playerOne") && playerOne.get()?.id != 0.0) {
+            val p1 = playerOne.get() ?: TTTGame_PlayerOne()
+            p1.id = 0.0
+            playerOne.set(p1)
+        }
+        if (handle.name.equals("playerTwo") && playerTwo.get()?.id != 1.0) {
+            val p2 = playerTwo.get() ?: TTTGame_PlayerOne()
+            p2.id = 1.0
+            playerTwo.set(p2)
+        }
+    }
+
+    override fun populateModel(slotName: String, model: Map<String, Any?>): Map<String, Any?> {
+        val gs = gameState.get() ?: TTTGame_GameState()
+
+        val cp = gs.currentPlayer ?: -1.0
+        val p1 = playerOne.get() ?: TTTGame_PlayerOne()
+        val p2 = playerTwo.get() ?: TTTGame_PlayerTwo()
+
+        val cpName = if (cp == p1.id) p1.name else p2.name
+        val cpAvatar = if (cp == p1.id) p1.avatar else p2.avatar
+
+        val winnerAvatar = gs.winnerAvatar ?: ""
+        val winnerName = if (winnerAvatar == p1.avatar) p1.name else p2.name
+        val congratsMessage = if (gs.winnerAvatar != null)
+            "Congratulations $winnerName, you won!" else "It's a tie!"
+
+        return model + mapOf(
+            "message" to if (gs.gameOver ?: false) congratsMessage else "",
+            "playerDetails" to "$cpName playing as $cpAvatar"
+        )
+    }
+
+    override fun onHandleUpdate(handle: Handle) {
+        val gs = gameState.get() ?: TTTGame_GameState()
+        // Apply the moves
+        if (!(gs.gameOver ?: false)) {
+            val board = gs.board ?: defaultGame.board!!
+            val boardList = board.split(",").toMutableList()
+            // Check the handle updated matches the current player
+            if (handle.name.equals("playerOneMove") && gs.currentPlayer == 0.0) {
+                applyMove(
+                    mv = playerOneMove.get()?.move?.toInt() ?: -1,
```

Comment:
Oh man, this'll be so much nicer when we have an updated kotlin api.

Reply:
10/10 would agree.